src/mpid/ch4/src/ch4_spawn.c

```diff
@@ -291,16 +293,25 @@ static int peer_intercomm_create(char *remote_addrname, int len, int tag,
     int mpi_errno = MPI_SUCCESS;
     int context_id, recvcontext_id;
     MPIR_Lpid remote_lpid;
+    bool need_unlock = false;
 
     mpi_errno = MPIR_Get_contextid_sparse(MPIR_Process.comm_self, &recvcontext_id, FALSE);
     MPIR_ERR_CHECK(mpi_errno);
 
+    /* We enter the LOCK to ensure the dynamic exchange don't get interleaved.
+     * NOTE: most other functions enter lock at NM-layer except the the functions defined
+     *       in e.g. ofi_spawn.c and ucx_spawn.c. So only those functions are allowed
+     *       inside the CS.
+     */
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI_LOCK(0));
+    need_unlock = true;
```

Comment:
was the previous implementation buggy or is this just moving the lock up a layer to be consistent across netmods?

Reply:
I believe the previous implementation is buggy. The dynamic connection is protected by a tag. Potentially multiple threads may connect to the same port and use the same tag.

The rest of the functions in `ofi_spawn.c` and `ucx_spawn.c` moved the lock for consistency. It is easier to track a rule file-wise.